qcodes/dataset/sqlite_base.py

```diff
@@ -209,10 +236,88 @@ def connect(name: str, debug: bool = False) -> sqlite3.Connection:
 
     if debug:
         conn.set_trace_callback(print)
+
+    init_db(conn)
+    perform_db_upgrade(conn, version=version)
     return conn
 
 
-def transaction(conn: sqlite3.Connection,
+def perform_db_upgrade(conn: SomeConnection, version: int=-1) -> None:
+    """
+    This is intended to perform all upgrades as needed to bring the
+    db from version 0 to the most current version (or the version specified).
+    All the perform_db_upgrade_X_to_Y functions most succeed or raise.
+
+    Args:
+        version: Which version to upgrade to. We count from 0. -1 means
+          'newest version'
+    """
+
+    upgrade_actions = {0: (lambda x: None,),
+                       1: (perform_db_upgrade_0_to_1,)}
+    newest_version = max(upgrade_actions.keys())
+    version = newest_version if version == -1 else version
+
+    current_version = get_user_version(conn)
+    if current_version < newest_version:
+        log.info("Commencing database upgrade")
+        for action in upgrade_actions[version]:
```

Comment:
the upgrades need to be executed in the correct order, right? do we rely here on the developers to list `upgrade_actions` in correct order from `0`? If so, please, add a comment to `upgrade_actions` about it. Otherwise, we need to sort ascending before looping, right?

Reply:
Well, the iteration is over the tuple that is the value of `upgrade_actions` for the key `version`. I can add a comment that `2` must have `(perform_db_upgrade_0_to_1, perform_db_upgrade_1_to_2)` if that is not clear from the context.

Reply:
oh, sorry, i missed the tuple part...

then the order within the tuple matters, right? i think this should be documented in the comments at least.

I feel that it is annoying and redundant to add two methods `(perform_db_upgrade_0_to_1, perform_db_upgrade_1_to_2)` for an upgrade to `2`. I also do not expect that for moving from `0` to `2` you'll need the two mentioned methods, but for moving from `0` to `3`, you'll need a different list of update functions, right?

I initially thought that each key has a single function assigned, which makes the upgrade from the *previous* version. This is actually also reflected in how the functions are currently called. The one-key-one-function approach also reflects perfectly the design: that all the upgrades shall be made sequentially, for example, an upgrade from 0 to 2 implies an upgrade from 0 to 1 and then from 1 to 2. What is the idea behind the dict with tuples?